// @microsoft/sp-property-pane/lib-commonjs/propertyPaneThumbnailPickerShared/ThumbnailPickerUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getThumbnailPickerButtons = void 0;
var tslib_1 = require("tslib");
var React = tslib_1.__importStar(require("react"));
var office_ui_fabric_react_bundle_1 = require("@ms/office-ui-fabric-react-bundle");
var PropertyPaneThumbnailStrings_resx_1 = tslib_1.__importDefault(require("./PropertyPaneThumbnailStrings.resx"));
var KillSwitches_1 = require("../common/KillSwitches");
function getThumbnailPickerButtons(handleChangeImageClick, onItemDeleted, buttonClassName, automationIdPrefix) {
    var changeButtonProps = {
        className: buttonClassName,
        onClick: handleChangeImageClick,
        ariaLabel: PropertyPaneThumbnailStrings_resx_1.default.PropertyPaneChangeImageButtonAriaLabel,
        ariaDescription: PropertyPaneThumbnailStrings_resx_1.default.PropertyPaneChangeImageButtonDescription,
        'data-automation-id': automationIdPrefix
            ? "".concat(automationIdPrefix, "ChangeButton")
            : 'thumbnailImagePickerChangeButton'
    };
    var changeButton = !(0, KillSwitches_1.isImageDeleteButtonKSActive)() ? (React.createElement(office_ui_fabric_react_bundle_1.ActionButton, tslib_1.__assign({}, tslib_1.__assign(tslib_1.__assign({}, changeButtonProps), { iconProps: {
            iconName: 'Edit'
        } })), PropertyPaneThumbnailStrings_resx_1.default.PropertyPaneChangeButtonLabel)) : (React.createElement(office_ui_fabric_react_bundle_1.PrimaryButton, tslib_1.__assign({}, changeButtonProps), PropertyPaneThumbnailStrings_resx_1.default.PropertyPaneChangeButtonLabel));
    var deleteButton = !(0, KillSwitches_1.isImageDeleteButtonKSActive)() && onItemDeleted ? (React.createElement(office_ui_fabric_react_bundle_1.ActionButton, { iconProps: { iconName: 'Delete' }, className: buttonClassName, onClick: onItemDeleted, ariaLabel: PropertyPaneThumbnailStrings_resx_1.default.PropertyPaneDeleteImageButtonAriaLabel, ariaDescription: PropertyPaneThumbnailStrings_resx_1.default.PropertyPaneDeleteImageButtonDescription, "data-automation-id": automationIdPrefix ? "".concat(automationIdPrefix, "DeleteButton") : 'thumbnailImagePickerDeleteButton' }, PropertyPaneThumbnailStrings_resx_1.default.PropertyPaneDeleteButtonLabel)) : undefined;
    if (deleteButton) {
        return (React.createElement(office_ui_fabric_react_bundle_1.Stack, { tokens: { childrenGap: 40 }, horizontal: true },
            changeButton,
            " ",
            deleteButton));
    }
    return changeButton;
}
exports.getThumbnailPickerButtons = getThumbnailPickerButtons;
