// @microsoft/sp-dialog/lib-commonjs/FabricDialogWrapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var Dialog_1 = require("@fluentui/react/lib/Dialog");
var Modal_1 = require("@fluentui/react/lib/Modal");
var PlaceholderSpinner_1 = tslib_1.__importDefault(require("./placeHolder/PlaceholderSpinner"));
var React = tslib_1.__importStar(require("react"));
var ContentDiv_1 = tslib_1.__importDefault(require("./ContentDiv"));
var DialogManager_1 = require("./DialogManager");
var Killswitches_1 = require("./common/Killswitches");
var FabricDialogWrapper = /** @class */ (function (_super) {
    tslib_1.__extends(FabricDialogWrapper, _super);
    function FabricDialogWrapper(props) {
        var _this = _super.call(this, props) || this;
        /**
         * Starts the sequence of opening the modal and rendering the dialog content into it.
         * Returns a promise that resolves when the content is fully rendered.
         */
        _this.open = function (dialog) {
            _this._openPromise = new Promise(function (res, rej) {
                _this._resolveOpenPromise = res;
            });
            dialog = dialog || _this.state.dialog;
            _this.setState({ dialog: dialog, isOpen: true, isContentReady: false });
            return _this._openPromise;
        };
        _this.close = function () {
            _this._closedPromise = new Promise(function (res, rej) {
                _this._resolveClosedPromise = res;
            });
            if (!_this.state.isOpen) {
                _this._resolveClosedPromise();
            }
            else {
                _this.setState({ isOpen: false });
            }
            return _this._closedPromise;
        };
        _this._onDismissed = function () {
            if (!_this.state.isOpen && _this._resolveClosedPromise) {
                _this._resolveClosedPromise();
            }
        };
        _this._closeCallback = function () {
            if (_this.props.closeCallback && _this.state.dialog) {
                _this.props.closeCallback(_this.state.dialog);
            }
        };
        _this._onDismissed = _this._onDismissed.bind(_this);
        _this.state = {
            isOpen: false,
            isContentReady: false
        };
        return _this;
    }
    FabricDialogWrapper.prototype.render = function () {
        var dialogContent = !this.state.isContentReady
            ? React.createElement(Dialog_1.DialogContent, {}, React.createElement(PlaceholderSpinner_1.default, { label: '' }))
            : false;
        if (this.state.dialog) {
            return (React.createElement(Modal_1.Modal, { isOpen: this.state.isOpen, isBlocking: this.state.dialog._configuration.isBlocking, layerProps: !(0, Killswitches_1.isEventBubblingEnabledKSActivated)()
                    ? {
                        eventBubblingEnabled: true
                    }
                    : undefined, onDismissed: this._onDismissed, onDismiss: this._closeCallback },
                ((0, Killswitches_1.isHandleMountingOfContentKSActivated)() || !dialogContent) && (React.createElement(ContentDiv_1.default, { innerElement: this._detachedElement })),
                dialogContent));
        }
        else {
            return React.createElement("div", null);
        }
    };
    FabricDialogWrapper.prototype.UNSAFE_componentWillMount = function () {
        this._detachedElement = document.createElement('DIV');
    };
    FabricDialogWrapper.prototype.componentDidUpdate = function () {
        var _this = this;
        if (this.state.isOpen && this.state.dialog && this._detachedElement) {
            if (!this.state.isContentReady) {
                (0, DialogManager_1._getInternalDialogApi)(this.state.dialog)
                    ._render(this._detachedElement)
                    .then(function () {
                    _this.setState({ isContentReady: true });
                });
            }
            else if (this._resolveOpenPromise) {
                this._resolveOpenPromise();
            }
        }
    };
    return FabricDialogWrapper;
}(React.Component));
exports.default = FabricDialogWrapper;
