// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Avatar/avatarSizes.js
"use strict";

exports.__esModule = true;
exports.statusSizeToPxValue = exports.statusIconSizeToPxValue = exports.sizeToPxValue = exports.iconSizeToPxValue = exports.getSizeStyles = void 0;
var _utils = require("../../../../utils");
/** Sizes for the Avatar and AvatarImage */
var sizeToPxValue = {
  smallest: 20,
  smaller: 24,
  small: 28,
  medium: 32,
  'medium-large': 36,
  large: 44,
  larger: 64,
  largest: 96
};

/** Sizes for the Icon inside the Avatar */
exports.sizeToPxValue = sizeToPxValue;
var iconSizeToPxValue = {
  smallest: 10,
  smaller: 12,
  small: 16,
  medium: 16,
  'medium-large': 20,
  large: 20,
  larger: 32,
  largest: 40
};

/** Sizes for the AvatarStatus and AvatarStatusImage */
exports.iconSizeToPxValue = iconSizeToPxValue;
var statusSizeToPxValue = {
  smallest: 6,
  smaller: 10,
  small: 10,
  medium: 10,
  'medium-large': 10,
  large: 10,
  larger: 16,
  largest: 0
};

/** Sizes for the AvaterStatusIcon */
exports.statusSizeToPxValue = statusSizeToPxValue;
var statusIconSizeToPxValue = {
  smallest: 4,
  smaller: 6,
  small: 6,
  medium: 6,
  'medium-large': 6,
  large: 6,
  larger: 10,
  largest: 0
};
exports.statusIconSizeToPxValue = statusIconSizeToPxValue;
var getSizeStyles = function getSizeStyles(sizeInPx) {
  var sizeInRem = (0, _utils.pxToRem)(sizeInPx);
  return {
    height: sizeInRem,
    width: sizeInRem
  };
};
exports.getSizeStyles = getSizeStyles;
