// @fluentui/react-northstar/dist/commonjs/themes/teams/components/Avatar/avatarStatusStyles.js
"use strict";

exports.__esModule = true;
exports.avatarStatusStyles = void 0;
var _avatarSizes = require("./avatarSizes");
var avatarStatusStyles = {
  root: function root(_ref) {
    var _ref$props = _ref.props,
      color = _ref$props.color,
      size = _ref$props.size,
      state = _ref$props.state,
      v = _ref.variables;
    return Object.assign({
      display: 'inline-flex',
      alignItems: 'center',
      justifyContent: 'center'
    }, (0, _avatarSizes.getSizeStyles)(_avatarSizes.statusSizeToPxValue[size]), {
      verticalAlign: 'middle',
      borderRadius: '9999px',
      position: 'absolute',
      bottom: 0,
      right: 0,
      boxShadow: "0 0 0 " + v.statusBorderWidth + " " + v.statusBorderColor
    }, state === 'success' && {
      backgroundColor: v.statusSuccessBackgroundColor
    }, state === 'info' && {
      backgroundColor: v.statusInfoBackgroundColor
    }, state === 'warning' && {
      backgroundColor: v.statusWarningBackgroundColor
    }, state === 'error' && {
      backgroundColor: v.statusErrorBackgroundColor
    }, state === 'unknown' && {
      backgroundColor: v.statusBackgroundColor
    }, !!color && {
      backgroundColor: color
    });
  }
};
exports.avatarStatusStyles = avatarStatusStyles;
