// @fluentui/accessibility/dist/commonjs/behaviors/Pill/pillBehavior.js
"use strict";

exports.__esModule = true;
exports.pillBehavior = void 0;
var _keyboardKey = require("../../keyboard-key");
var _attributes = require("../../attributes");
var pillBehavior = function pillBehavior(p) {
  var _Object$assign;
  return {
    attributes: {
      root: Object.assign((_Object$assign = {
        role: p.actionable || p.selectable ? 'button' : 'none',
        tabIndex: p.actionable || p.selectable ? 0 : -1
      }, _Object$assign[_attributes.IS_FOCUSABLE_ATTRIBUTE] = p.actionable || p.selectable || p.role === 'option', _Object$assign), p.selectable && {
        'aria-selected': p.selected
      })
    },
    keyActions: {
      root: Object.assign({}, p.dismissible && {
        performDismiss: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.Delete
          }, {
            keyCode: _keyboardKey.keyboardKey.Backspace
          }]
        }
      }, (p.selectable || p.actionable) && {
        performClick: {
          keyCombinations: [{
            keyCode: _keyboardKey.keyboardKey.Enter
          }, {
            keyCode: _keyboardKey.SpacebarKey
          }]
        }
      })
    }
  };
};
exports.pillBehavior = pillBehavior;
