// @microsoft/sp-listview-extensibility/lib-commonjs/common/ListViewAccessorState.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListViewAccessorStateChanges = void 0;
/**
 * Describes the type of list view state changes.
 *
 * @public
 */
var ListViewAccessorStateChanges;
(function (ListViewAccessorStateChanges) {
    ListViewAccessorStateChanges[ListViewAccessorStateChanges["None"] = 0] = "None";
    ListViewAccessorStateChanges[ListViewAccessorStateChanges["List"] = 1] = "List";
    ListViewAccessorStateChanges[ListViewAccessorStateChanges["View"] = 2] = "View";
    ListViewAccessorStateChanges[ListViewAccessorStateChanges["Columns"] = 4] = "Columns";
    ListViewAccessorStateChanges[ListViewAccessorStateChanges["SelectedRows"] = 8] = "SelectedRows";
    ListViewAccessorStateChanges[ListViewAccessorStateChanges["Rows"] = 16] = "Rows";
    ListViewAccessorStateChanges[ListViewAccessorStateChanges["AppliedFilters"] = 32] = "AppliedFilters";
    ListViewAccessorStateChanges[ListViewAccessorStateChanges["Sort"] = 64] = "Sort";
    ListViewAccessorStateChanges[ListViewAccessorStateChanges["FolderInfo"] = 128] = "FolderInfo";
})(ListViewAccessorStateChanges = exports.ListViewAccessorStateChanges || (exports.ListViewAccessorStateChanges = {}));
