// @microsoft/sp-application-base/lib-commonjs/frameworkPlaceholders/AadPlaceholderManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var sp_http_base_1 = require("@microsoft/sp-http-base");
var Killswitches_1 = require("../common/Killswitches");
var SPApplicationBase_resx_1 = tslib_1.__importDefault(require("../SPApplicationBase.resx"));
var PlaceholderProvider_1 = tslib_1.__importDefault(require("../extensibility/placeholder/PlaceholderProvider"));
var PlaceholderName_1 = tslib_1.__importDefault(require("../extensibility/placeholder/PlaceholderName"));
var PlaceholderManager_1 = tslib_1.__importDefault(require("../extensibility/placeholder/PlaceholderManager"));
/**
 * This class sets up an a placeholder for scenarios that token acquisition fails and will require
 * end user interaction.
 * @internal
 */
var AadPlaceholderManager = /** @class */ (function () {
    function AadPlaceholderManager() {
        this._didRenderError = false;
    }
    /**
     * Sets up a placeholder (or alert if unavailable) for the token acquisition failure event.
     */
    AadPlaceholderManager.prototype.setUpTokenAcquistionFailurePlaceholder = function (application, serviceScope) {
        var _this = this;
        var tokenProvider = sp_http_base_1._AadTokenProviders.configurable;
        var placeHolderProvider = new PlaceholderProvider_1.default(serviceScope, {
            sequence: 0,
            preAllocatedApplicationCustomizerBottomHeight: 0,
            preAllocatedApplicationCustomizerTopHeight: 0
        }, 'AAD Failure Placeholder'
        /* This placeholder should render over any other placeholder */
        );
        var placeholderManager = serviceScope.consume(PlaceholderManager_1.default.serviceKey);
        placeHolderProvider.changedEvent.add(application, function () {
            if (!_this._aadPlaceHolderContent) {
                _this._aadPlaceHolderContent = placeHolderProvider.tryCreateContent(PlaceholderName_1.default.Top);
            }
        });
        tokenProvider.tokenAcquisitionEvent.add(application, function (eventArgs) {
            if (!_this._didRenderError) {
                _this.renderTokenAcquistionEvent(placeholderManager.isEnabled, eventArgs);
            }
        });
    };
    AadPlaceholderManager.prototype.renderTokenAcquistionEvent = function (isPlaceholderManagerEnabled, eventArgs) {
        if (Killswitches_1.Killswitches.isDuplicateErrorMessageV2KillSwitchActivated() || !this._didRenderError) {
            if (!isPlaceholderManagerEnabled) {
                this._renderTokenAcquistionAlert(eventArgs);
            }
            else {
                if (this._aadPlaceHolderContent) {
                    this._renderTokenAcquistionPlaceholder(this._aadPlaceHolderContent, eventArgs);
                }
                else {
                    this._renderTokenAcquistionAlert(eventArgs);
                }
            }
        }
        this._didRenderError = true;
    };
    AadPlaceholderManager.prototype._renderTokenAcquistionPlaceholder = function (placeholderContent, eventArgs) {
        placeholderContent.domElement.setAttribute('style', 'text-align: center; padding: 10px; background-color:#fff4ce;');
        var errorMessageElement = document.createElement('span');
        errorMessageElement.innerText = eventArgs.message;
        placeholderContent.domElement.appendChild(errorMessageElement);
        if (eventArgs.redirectUrl) {
            var resolveLinkElement = document.createElement('a');
            resolveLinkElement.href = eventArgs.redirectUrl;
            resolveLinkElement.innerText = SPApplicationBase_resx_1.default.clickToResolveIssueLinkText;
            placeholderContent.domElement.appendChild(resolveLinkElement);
        }
    };
    AadPlaceholderManager.prototype._renderTokenAcquistionAlert = function (eventArgs) {
        if (confirm(eventArgs.message) && eventArgs.redirectUrl) {
            window.location.href = eventArgs.redirectUrl;
        }
    };
    return AadPlaceholderManager;
}());
exports.default = AadPlaceholderManager;
