// @fluentui/react-northstar/dist/dts/src/components/Avatar/AvatarStatusImage.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AvatarStatusImage = exports.avatarStatusImageClassName = void 0;
var React = require("react");
var accessibility_1 = require("@fluentui/accessibility");
var react_bindings_1 = require("@fluentui/react-bindings");
var customPropTypes = require("@fluentui/react-proptypes");
var utils_1 = require("../../utils");
exports.avatarStatusImageClassName = 'ui-avatar__statusimage';
/**
 * A AvatarStatusImage provides a status image for the Avatar.
 */
exports.AvatarStatusImage = React.forwardRef(function (props, ref) {
    var context = react_bindings_1.useFluentContext();
    var _a = react_bindings_1.useTelemetry(exports.AvatarStatusImage.displayName, context.telemetry), setStart = _a.setStart, setEnd = _a.setEnd;
    setStart();
    var children = props.children, className = props.className, design = props.design, size = props.size, styles = props.styles, variables = props.variables;
    var classes = react_bindings_1.useStyles(exports.AvatarStatusImage.displayName, {
        className: exports.avatarStatusImageClassName,
        mapPropsToStyles: function () { return ({
            size: size,
        }); },
        mapPropsToInlineStyles: function () { return ({
            className: className,
            design: design,
            styles: styles,
            variables: variables,
        }); },
        rtl: context.rtl,
    }).classes;
    var getA11Props = react_bindings_1.useAccessibility(props.accessibility, {
        debugName: exports.AvatarStatusImage.displayName,
        rtl: context.rtl,
    });
    var ElementType = react_bindings_1.getElementType(props);
    var unhandledProps = react_bindings_1.useUnhandledProps(exports.AvatarStatusImage.handledProps, props);
    var element = (React.createElement(ElementType, __assign({}, getA11Props('root', __assign({ className: classes.root, ref: ref }, unhandledProps))), children));
    setEnd();
    return element;
});
exports.AvatarStatusImage.displayName = 'AvatarStatusImage';
exports.AvatarStatusImage.propTypes = __assign(__assign({}, utils_1.commonPropTypes.createCommon()), { size: customPropTypes.size });
exports.AvatarStatusImage.handledProps = Object.keys(exports.AvatarStatusImage.propTypes);
exports.AvatarStatusImage.defaultProps = {
    accessibility: accessibility_1.imageBehavior,
    as: 'img',
};
exports.AvatarStatusImage.shorthandConfig = {
    allowsJSX: false,
    mappedProp: 'src',
};
exports.AvatarStatusImage.create = utils_1.createShorthandFactory({
    allowsJSX: false,
    Component: exports.AvatarStatusImage,
    mappedProp: 'src',
});
