// @fluentui/react-northstar/dist/es/themes/teams/components/Loader/loaderVariables.js
import { pxToRem } from '../../../../utils';
export var loaderVariables = function loaderVariables(siteVariables) {
  return {
    containerHeights: {
      smallest: pxToRem(24),
      smaller: pxToRem(24),
      small: pxToRem(24),
      medium: pxToRem(36),
      large: pxToRem(72),
      larger: pxToRem(72),
      largest: pxToRem(72)
    },
    containerWidths: {
      smallest: pxToRem(24),
      smaller: pxToRem(24),
      small: pxToRem(24),
      medium: pxToRem(36),
      large: pxToRem(72),
      larger: pxToRem(72),
      largest: pxToRem(72)
    },
    svgHeights: {
      smallest: pxToRem(24),
      smaller: pxToRem(24),
      small: pxToRem(24),
      medium: pxToRem(36),
      large: pxToRem(72),
      larger: pxToRem(72),
      largest: pxToRem(72)
    },
    svgTranslatePosition: {
      smallest: pxToRem(-1440),
      smaller: pxToRem(-1440),
      small: pxToRem(-1440),
      medium: pxToRem(-2160),
      large: pxToRem(-4320),
      larger: pxToRem(-4320),
      largest: pxToRem(-4320)
    },
    svgWidths: {
      smallest: pxToRem(24),
      smaller: pxToRem(24),
      small: pxToRem(24),
      medium: pxToRem(36),
      large: pxToRem(72),
      larger: pxToRem(72),
      largest: pxToRem(72)
    },
    svgTrackColor: siteVariables.colorScheme.brand.border2,
    svgTailColor: siteVariables.colorScheme.brand.foreground,
    svgSecondaryColor: siteVariables.colorScheme.default.foreground3
  };
};
