// @fluentui/react-northstar/dist/dts/src/themes/teams/components/Loader/loaderStyles.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.loaderStyles = void 0;
var utils_1 = require("../../../../utils");
var rootFlexDirections = {
    above: 'column-reverse',
    below: 'column',
    start: 'row-reverse',
    end: 'row',
};
var getCircleStyles = function () { return ({
    cx: '50%',
    cy: '50%',
    r: '45%',
    fill: 'none',
    strokeWidth: '8px',
}); };
exports.loaderStyles = {
    root: function (_a) {
        var p = _a.props;
        return ({
            alignItems: 'center',
            display: p.inline ? 'inline-flex' : 'flex',
            justifyContent: 'center',
            flexDirection: rootFlexDirections[p.labelPosition],
        });
    },
    indicator: function (_a) {
        var p = _a.props, v = _a.variables;
        return ({
            height: v.containerHeights[p.size],
            width: v.containerWidths[p.size],
        });
    },
    svg: function (_a) {
        var p = _a.props, v = _a.variables;
        return ({
            animation: '3s linear infinite',
            animationName: {
                '0%': {
                    transform: 'rotate(0deg) /* @noflip */',
                },
                '100%': {
                    transform: 'rotate(360deg) /* @noflip */',
                },
            },
            width: v.svgWidths[p.size],
            height: v.svgHeights[p.size],
        });
    },
    svgTrack: function (_a) {
        var p = _a.props, v = _a.variables;
        return (__assign({ stroke: !p.secondary && v.svgTrackColor }, getCircleStyles()));
    },
    svgTail: function (_a) {
        var p = _a.props, v = _a.variables;
        return (__assign(__assign({}, getCircleStyles()), { animation: '1.5s cubic-bezier(0.33,0,0.67,1) infinite', strokeLinecap: 'round', strokeDasharray: '283', strokeDashoffset: '280', transformOrigin: '50% 50%', stroke: p.secondary ? v.svgSecondaryColor : v.svgTailColor, animationName: {
                '0%': {
                    strokeDashoffset: '283',
                    transform: 'rotate(0) /* @noflip */',
                },
                '25%': {
                    strokeDashoffset: '283',
                    transform: 'rotate(0) /* @noflip */',
                },
                '50%': {
                    strokeDashoffset: '75',
                    transform: 'rotate(45deg) /* @noflip */',
                },
                '75%': {
                    strokeDashoffset: '75',
                    transform: 'rotate(45deg) /* @noflip */',
                },
                '100%': {
                    strokeDashoffset: '283',
                    transform: 'rotate(360deg) /* @noflip */',
                },
            } }));
    },
    label: function () { return ({
        margin: utils_1.pxToRem(10),
    }); },
};
