// @fluentui/react-northstar-fela-renderer/dist/commonjs/felaExpandCssShorthandsPlugin.js
"use strict";

exports.__esModule = true;
exports.felaExpandCssShorthandsPlugin = void 0;
var _inlineStyleExpandShorthand = require("inline-style-expand-shorthand");
// @ts-ignore

// https://jsperf.com/array-indexof-vs-object-key-lookup2/12
var handledCssProps = {
  // 'font', Oops, is not supported by inline-style-expand-shorthand
  padding: true,
  margin: true,
  border: true,
  borderWidth: true,
  borderStyle: true,
  borderColor: true,
  borderTop: true,
  borderRight: true,
  borderBottom: true,
  borderLeft: true,
  borderRadius: true,
  background: true,
  outline: true
};
var felaExpandCssShorthandsPlugin = function felaExpandCssShorthandsPlugin(styles) {
  return Object.keys(styles).reduce(function (acc, cssPropertyName) {
    var cssPropertyValue = styles[cssPropertyName];
    if (cssPropertyValue === null || typeof cssPropertyValue === 'undefined') {
      acc[cssPropertyName] = cssPropertyValue;
    } else if (handledCssProps[cssPropertyName]) {
      var expandedProps = (0, _inlineStyleExpandShorthand.expandProperty)(cssPropertyName, cssPropertyValue);
      if (expandedProps) {
        Object.assign(acc, expandedProps);
        return acc;
      }
      acc[cssPropertyName] = cssPropertyValue;
    } else if (Array.isArray(cssPropertyValue)) {
      acc[cssPropertyName] = cssPropertyValue;
    } else if (typeof cssPropertyValue === 'object') {
      acc[cssPropertyName] = felaExpandCssShorthandsPlugin(cssPropertyValue);
    } else {
      acc[cssPropertyName] = cssPropertyValue;
    }
    return acc;
  }, {});
};
exports.felaExpandCssShorthandsPlugin = felaExpandCssShorthandsPlugin;
